lib/fd.js

```diff
@@ -0,0 +1,309 @@
+      if (err) return cb(err)
+      if (blkOffset) data = data.slice(blkOffset)
+
+      data.copy(buf)
+      const read = Math.min(data.length, buf.length)
+
+      if (blk === this.blockPosition && blkOffset === this.blockOffset) {
+        this.position += read
+        if (read === data.length) {
+          this.blockPosition++
+          this.blockOffset = 0
+        } else {
+          this.blockOffset = blkOffset + read
+        }
+      }
+
+      cb(null, read, buffer)
+    })
+  }
+}
+
+FileDescriptor.create = function (drive, name, flags, cb) {
+  try {
+    flags = toFlagsNumber(flags)
+  } catch (err) {
+    return cb(err)
+  }
+
+  const accmode = flags & O_ACCMODE
+  const writable = !!(accmode & (O_WRONLY | O_RDWR))
+  const readable = accmode === 0 || !!(accmode & O_RDWR)
+  const appending = !!(flags & O_APPEND)
+  const creating = !!(flags & O_CREAT)
+  const canExist = !(flags & O_EXCL)
+
+  drive.contentReady(err => {
+    if (err) return cb(err)
+    drive._db.get(name, (err, st) => {
+      if (err) return cb(err)
+      if (st && !canExist) return cb(new errors.PathAlreadyExists(name))
+      if (!st && (!writable || !creating)) return cb(new errors.FileNotFound(name))
+
+      if (st) {
+        try { 
+          st = messages.Stat.decode(st.value)
+        } catch (err) {
+          return cb(err)
+        }
+      }
+
+      const fd = new FileDescriptor(drive, name, st, readable, writable, appending, creating)
+      if (creating) {
+        drive.create(name, err => {
+          if (err) return cb(err)
+          return cb(null, fd)
+        })
+      } else {
+        return cb(null, fd)
+      }
+    })
+  })
+}
+
+module.exports = FileDescriptor
+
+// Copied from the Node FS internal utils.
+function toFlagsNumber (flags) {
+  if (typeof flags === 'number') {
+    return flags
+  }
+
+  switch (flags) {
+    case 'r' : return O_RDONLY
+    case 'rs' : // Fall through.
+    case 'sr' : return O_RDONLY | O_SYNC
+    case 'r+' : return O_RDWR
+    case 'rs+' : // Fall through.
+    case 'sr+' : return O_RDWR | O_SYNC
+
+    case 'w' : return O_TRUNC | O_CREAT | O_WRONLY
+    case 'wx' : // Fall through.
+    case 'xw' : return O_TRUNC | O_CREAT | O_WRONLY | O_EXCL
+
+    case 'w+' : return O_TRUNC | O_CREAT | O_RDWR
+    case 'wx+': // Fall through.
+    case 'xw+': return O_TRUNC | O_CREAT | O_RDWR | O_EXCL
+
+    case 'a' : return O_APPEND | O_CREAT | O_WRONLY
+    case 'ax' : // Fall through.
+    case 'xa' : return O_APPEND | O_CREAT | O_WRONLY | O_EXCL
+    case 'as' : // Fall through.
+    case 'sa' : return O_APPEND | O_CREAT | O_WRONLY | O_SYNC
+
+    case 'a+' : return O_APPEND | O_CREAT | O_RDWR
+    case 'ax+': // Fall through.
+    case 'xa+': return O_APPEND | O_CREAT | O_RDWR | O_EXCL
+    case 'as+': // Fall through.
+    case 'sa+': return O_APPEND | O_CREAT | O_RDWR | O_SYNC
+  }
+
+  throw new errors.InvalidArgument(`Invalid value in flags: ${flags}`)
+}
```

Comment:
the modularity in me screams make this func a separate module! :D :D :D

Reply:
Done!